docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,3025 @@ p4grammar {
+|   | $\vert$ | $\expMem \exp \field$ | $\qquad \text{expression members}$ |
+|   | $\vert$ | $\ternary {\exp_1} {\exp_2} {\exp_3}$ | $\qquad \text{conditionals}$ |
+|   | $\vert$ | $\funcCall \exp {\baseTyps} \args$ | $\qquad \text{function calls}$ |
+|   | $\vert$ | $\instantiation {\baseTyp} {\args}$ | $\qquad \text{anonymous instantiation}$ |
+|   | $\vert$ | $\mask {\exp_1} {\exp_2}$ | $\qquad \text{bit masks}$ |
+|   | $\vert$ | $\range {\exp_1} {\exp_2}$ | $\qquad \text{ranges}$ |
+{ .booktable }
+
+**Labels:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\lbl$ | $::=$ | $\defLbl$ | $\qquad \text{default}$ |
+|   | $\vert$ | $\str$ | $\qquad \text{string labels}$ |
+{ .booktable }
+
+**Switchs:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\switchCase$ | $::=$ | $\actionCase \stmts$ | $\qquad \text{labeled block}$ |
+|   | $\vert$ | $\lbl$ | $\qquad \text{fall through with label}$ |
+{ .booktable }
+
+**Statements:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\stmt$ | $::=$ | $\methodCall \exp \baseTyps \args$ | $\qquad \text{method call}$ |
+|   | $\vert$ | $\assign {\exp_1} {\exp_2}$ | $\qquad \text{assignment}$ |
+|   | $\vert$ | $\dirApp \baseTyp \args$ | $\qquad \text{direct application}$ |
+|   | $\vert$ | $\ifthen \exp \stmt$ | $\qquad \text{if then}$ |
+|   | $\vert$ | $\ifthenelse \exp {\stmt_1} {\stmt_2}$ | $\qquad \text{if then else}$ |
+|   | $\vert$ | $\block \stmts$ | $\qquad \text{sequencing}$ |
+|   | $\vert$ | $\exit$ | $\qquad \text{exit}$ |
+|   | $\vert$ | $\noop$ | $\qquad \text{noop}$ |
+|   | $\vert$ | $\return \exp$ | $\qquad \text{return}$ |
+|   | $\vert$ | $\retNothing$ | $\qquad \text{return nothing}$ |
+|   | $\vert$ | $\switch \exp \switchCases$ | $\qquad \text{switch}$ |
+|   | $\vert$ | $\dcl$ | $\qquad \text{declaration}$ |
+{ .booktable }
+
+**Parser states:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\nxt$ | $::=$ | $\str$ | $\qquad \text{next state}$ |
+| $\pmatch$ | $::=$ | $\defMatch$ | $\qquad \text{default match}$ |
+|   | $\vert$ | $\dontcareMatch$ | $\qquad \text{don't care match}$ |
+|   | $\vert$ | $\exp$ | $\qquad \text{expression match}$ |
+| $\case$ | $::=$ | $\expandCase \pmatches \nxt$ | $\qquad \text{match cases for the next state}$ |
+| $\parserTran$ | $::=$ | $\nxt$ | $\qquad \text{next state transition}$ |
+|   | $\vert$ | $\slctTran \exp \pcases$ | $\qquad \text{select next state transition}$ |
+| $\state$ | $::=$ | $\stateDef \var \stmts \parserTran$ | $\qquad \text{states}$ |
+{ .booktable }
+
+**Table properties:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\action$ | $::=$ | $\actionDef \var \arg$ | $\qquad \text{actions}$ |
+| $\prop$ | $::=$ | $\key \exp \var$ | $\qquad \text{keys}$ |
+|   | $\vert$ | $\action$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\entry \pmatch \action$ | $\qquad \text{entries}$ |
+|   | $\vert$ | $\custom \bool \var \exp$ | $\qquad \text{custom property}$ |
```

Comment:
what is the \bool here for?

Reply:
I think it determines if the custom property is constant or not. This was the version based on Petr4 and it's defined as (I may end up dropping it after revision to base the formalization on P4):

`  type 'a pproperty =
      Key of
        { tags: 'a;
          keys: key list }
      (** Keys *)
    | Actions of
        { tags: 'a;
          actions: action_ref list }
      (** Actions *)
    | Entries of
        { tags: 'a;
          entries: entry list }
      (** Entries *)
    | Custom of
        { tags: 'a;
          annotations: Annotation.t list;
          const: bool;
          name: P4String.t;
          value: Expression.t }
      (** Custom (e.g., [size], [default_action], etc.) *)
  [@@deriving sexp,show,yojson]
`

Reply:
I guess an alternative to `custom b x e` would be to have two variants `custom x e` and `custom const x e`, or something like that.